lib/tokens.ts

```diff
@@ -0,0 +1,191 @@
+import { PublicKey } from "@solana/web3.js";
+import { TokenProps } from "./types";
+import {
+  JsonMetadata,
+  MPL_TOKEN_METADATA_PROGRAM_ID,
+  deserializeMetadata,
+} from "@metaplex-foundation/mpl-token-metadata";
+import { RpcAccount } from "@metaplex-foundation/umi";
+import { Provider } from "@coral-xyz/anchor";
+import { USDCAddress, USDCMetadata } from "./constants";
+import {
+  Mint,
+  TOKEN_2022_PROGRAM_ID,
+  getMint,
+  getTokenMetadata,
+} from "@solana/spl-token";
+/**
+ * Starts with the jup.ag strict list to find token. jup.ag maintains a list of quality tokens
+ * if that fails, use metaplex with RPC call and fetch metadata json
+ * if THAT fails, take MINT address and truncate by 5 characters and that's the symbol (no image)
+ */
+export async function enrichTokenMetadata(
+  tokenAddress: PublicKey,
+  rpcProvider: Provider
+): Promise<TokenProps> {
+  // first, is it USDC?
+  if (tokenAddress.toString() === USDCAddress) {
+    return USDCMetadata;
+  }
+
+  // second check jup list
+  const tokenOnJup = await getTokenFromJupStrictList(tokenAddress);
+  if (tokenOnJup) {
+    return {
+      symbol: tokenOnJup.symbol,
+      publicKey: tokenAddress.toString(),
+      url: tokenOnJup.logoURI,
+      decimals: tokenOnJup.decimals,
+    };
+  }
+
+  // next, try metaplex
+  const jsonMetadata = await getMetaplexMetadataForToken(
+    tokenAddress,
+    rpcProvider
+  );
+  if (jsonMetadata && jsonMetadata.symbol) {
+    return {
+      symbol: jsonMetadata.symbol,
+      publicKey: tokenAddress.toString(),
+      url: jsonMetadata.image,
+      decimals: jsonMetadata.seller_fee_basis_points,
+    };
+  }
+  let mint: Mint | undefined = undefined;
+  try {
+    //try getting metadata from token 2022
+    const token2022Metadata = await getTokenMetadata(
```

Comment:
Theres ONE MORE you need to include which is just the regular tokenkeg TOKEN_PROGRAM_ID to check for. Sorry that's what I kept saying here, it's not just token-2022 as that isnt required to be a token, it's just that its the new one, but doesn't take over the old one.

So metaplex, token-2022, token, string.